spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapterTests.java

```diff
@@ -16,28 +16,171 @@
+import org.springframework.messaging.MessageChannel;
 
-import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 /**
  * {@link PubSubInboundChannelAdapter} unit tests.
  *
  * @author João André Martins
+ * @author Doug Hoard
  */
 @RunWith(MockitoJUnitRunner.class)
 public class PubSubInboundChannelAdapterTests {
 
-	@Mock
-	private PubSubTemplate pubSubTemplate;
+	public static final String ACK = "ACK";
+
+	public static final String NACK = "NACK";
+
+	private PubSubOperations pubSubOperations;
+
+	private PubSubSubscriberOperations pubSubSubscriberOperations;
+
+	private MessageChannel messageChannel;
+
+	private AtomicReference<String> returnValue;
+
+	private CountDownLatch countDownLatch;
+
+	private AckReplyConsumer ackReplyConsumer;
+
+	private TestAnswer testAnswer;
+
+	@Before
+	public void setUp() {
+		this.pubSubOperations = mock(PubSubOperations.class);
+		this.pubSubSubscriberOperations = mock(PubSubSubscriberOperations.class);
+		this.messageChannel = mock(MessageChannel.class);
+
+		when(this.messageChannel.send(any())).thenThrow(new RuntimeException("Forced exception sending message"));
+
+		this.returnValue = new AtomicReference<>();
+		this.countDownLatch = new CountDownLatch(1);
+
+		NackAnswer nackAnwser = new NackAnswer(this.returnValue);
+
+		this.ackReplyConsumer = mock(AckReplyConsumer.class);
+
+		doAnswer(nackAnwser).when(this.ackReplyConsumer).nack();
+
+		this.testAnswer = new TestAnswer(this.countDownLatch, this.ackReplyConsumer);
+	}
 
 	@Test(expected = IllegalArgumentException.class)
 	public void testNonNullAckMode() {
 		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
-				this.pubSubTemplate, "testSubscription");
+				this.pubSubOperations, "testSubscription");
+
 		adapter.setAckMode(null);
 	}
+
+	@Test
+	public void testAckModeAuto() throws Exception {
+		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
+				this.pubSubSubscriberOperations, "testSubscription");
+
+		adapter.setAckMode(AckMode.AUTO);
+
+		when(this.pubSubSubscriberOperations.subscribe(anyString(), any(MessageReceiver.class))).then(this.testAnswer);
+
+		adapter.setOutputChannel(this.messageChannel);
+
+		try {
+			adapter.start();
+		}
+		catch (Throwable t) {
+
+		}
+
+		Assert.assertTrue(this.countDownLatch.await(10, TimeUnit.SECONDS));
+		Assert.assertEquals(NACK, this.returnValue.get());
+	}
+
+	@Test
+	public void testAckModeAutoAck() throws Exception {
+		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
+				this.pubSubSubscriberOperations, "testSubscription");
+
+		adapter.setAckMode(AckMode.AUTO_ACK);
+
+		when(this.pubSubSubscriberOperations.subscribe(anyString(), any(MessageReceiver.class))).then(this.testAnswer);
```

Comment:
No, you can't stub the same method on the same mock object, independently of the place.
Although I see you expect and produce the same answer. So, would be great to move this stubbing in that `setUp()` - and all your test will still work well!